src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSyncController.scala

```diff
@@ -0,0 +1,171 @@
+  }
+
+  private def processBlockHeaders(headers: Seq[BlockHeader], peer: ActorRef) = {
+    val parentByNumber = blockchain.getBlockHeaderByNumber(headers.head.number - 1)
+
+    parentByNumber match {
+      case Some(parent) if checkHeaders(headers) =>
+        //we have same chain
+        if (parent.hash == headers.head.parentHash) {
+          peer ! SendMessage(GetBlockBodies(headers.take(blockBodiesPerRequest).map(_.hash)))
+        } else {
+          context.self ! ResolveBranch(sender())
+        }
+      case None =>
+        log.warning("got header that does not have parent")
+        resumeWithDifferentPeer()
+    }
+  }
+
+  private def handleBlockBodies(m: BlockBodies) = {
+    if (m.bodies.nonEmpty) {
+      val result = headers.zip(m.bodies).map { case (h, b) => blockValidator(h, b) }
+
+      if (!result.exists(_.isLeft)) {
+        val blocks = result.collect { case Right(b) => b }
+
+        blockchain.getBlockHeaderByHash(blocks.head.header.parentHash)
+          .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash)) match {
+          case Some(td) =>
+            var currentTd = td
+            val newBlocks = blocks.map { b =>
+              val blockHashToDelete = blockchain.getBlockHeaderByNumber(b.header.number).map(_.hash).filter(_ != b.header.hash)
+              blockchain.save(b)
+              appStateStorage.putBestBlockNumber(b.header.number)
+              currentTd += b.header.difficulty
+              blockchain.save(b.header.hash, currentTd)
+              (blockHashToDelete, NewBlock(b, currentTd))
+            }
+
+            newBlocks.collect { case (Some(hash), _) => hash }.foreach(blockchain.removeBlock)
+            context.self ! BroadcastBlocks(newBlocks.map(_._2))
+          case None =>
+            log.error("no total difficulty for latest block")
+        }
+
+        headers = headers.drop(result.length)
+        if (headers.nonEmpty) {
+          sender() ! SendMessage(GetBlockBodies(headers.take(blockBodiesPerRequest).map(_.hash)))
+        } else {
+          context.self ! ResumeRegularSync
+        }
+      } else {
+        //blacklist for errors in blocks
+        resumeWithDifferentPeer()
+      }
+    } else {
+      //we got empty response for bodies from peer but we got block headers earlier
+      resumeWithDifferentPeer()
+    }
+  }
+
+  private def scheduleResume = {
+    headers = Seq.empty
+    scheduler.scheduleOnce(checkForNewBlockInterval, context.self, ResumeRegularSync)
+  }
+
+  private def resumeWithDifferentPeer() = {
+    blacklist(sender(), blacklistDuration)
+    headers = Seq.empty
+    context.self ! ResumeRegularSync
+  }
+
+  private def checkHeaders(headers: Seq[BlockHeader]): Boolean =
+    headers.zip(headers.tail).forall { case (parent, child) => parent.hash == child.parentHash && parent.number + 1 == child.number }
+
+  private def bestPeer: Option[ActorRef] = Try(peersToDownloadFrom.maxBy { case (_, status) => status.totalDifficulty }._1).toOption
```

Comment:
just a note: this status is the initial status we received in `Status` message, it does not necessarily mean this peer is still the best one.